Improve SetParametersAsync docs

rynowak opened the pull request:
Fixes: #12820

Commit:
Improve SetParametersAsync docs

Fixes: #12820

rynowak commented:
@tstivers1990 - anything else you want to see added here?

tstivers1990 commented:
@rynowak That clarifies things perfectly. Thank you.

SteveSandersonMS commented on src/Components/Components/src/ComponentBase.cs:
```diff
@@ -171,10 +171,25 @@ void IComponent.Attach(RenderHandle renderHandle)
+        /// </para>
+        /// <para>
+        /// The default implementation of <see cref="SetParametersAsync(ParameterView)"/> will set a value for every property
+        /// decorated with <see cref="ParameterAttribute"/> or <see cref="CascadingParameterAttribute"/> each time
```
I'm a bit surprised by this. Maybe it's something we should consider locking down a bit more before 3.0 ships.

I was expecting that we only assign parameter values where `.Cascading == true` to `[CascadingParameter]` targets, and only where `.Cascading == false` to `[Parameter]` targets. I didn't think it was desired that you can overwrite a `[CascadingParameter]` by explicitly supplying a non-cascading value from the parent. The tooling implies you can't (no cascading parameters appear in intellisense) so if the runtime behaves differently from that it's surprising. The `Cascading` flag on `ParameterValue` is populated but apparently not read back anywhere, which seems very odd.

Do you agree? If so I think we could amend the logic in `ComponentProperties.cs` quite easily to behave this way.

rynowak replied:
> I'm a bit surprised by this. Maybe it's something we should consider locking down a bit more before 3.0 ships.

I'll try this out. I'm not trying to say something contraversial. I agree that this should be locked down.

Commit:
Clarify semantics of clashing cascading parameters

This change attempts to lock down the semantics of how cascading
parameters behave inside ParameterView.SetParameterProperties.

Previously a cascading value could be used to set a non-cascading
parameter, and vice-versa. We were just missing tests in general for how
cascading values are interpreted by the parameter set logic.

rynowak commented:
@SteveSandersonMS - added lots of tests and some improvements.

rynowak commented:
@SteveSandersonMS - ready for another look.

SteveSandersonMS commented on src/Components/Components/src/Reflection/ComponentProperties.cs:
```diff
@@ -44,6 +47,24 @@ public static void SetProperties(in ParameterView parameters, object target)
+                    {
+                        // We're giving a more specific error here because trying to set a non-cascading parameter
+                        // with a cascading value is likely deliberate (but not supported), or is a bug in our code.
+                        ThrowForSettingParameterWithCascadingValue(targetType, parameterName);
```
My understanding is that this can't happen by accident (e.g., there happens by coincidence to be a CascadingValue with the same name as a non-cascading parameter), because when setting up the set of subscriptions to cascading parameters, we only pick the ones that the component asks for.

As such this code path only occurs if there's some kind of bug in the renderer (i.e., it's supplying an unwanted cascading parameter).

Assuming that's all true this is good and fine, though it might be nice to clarify the comment on L63-64 to emphasise how if this happens, it's a bug in the renderer and not in user code.

SteveSandersonMS replied:
> it's a bug in the renderer and not in user code

Or highly nonstandard user code (passing a manually-constructed `ParameterView`), or more likely, user unit test code. Right?

rynowak replied:
Right, I'm referring to the case where a user constructed their own parameter view (`is likely deliberate (but not supported` ) whether it's in tests or not. 

I plan on leaving the comment as is - I think it's already sufficiently descriptive.

SteveSandersonMS commented on src/Components/Components/test/ParameterViewTest.Assignment.cs:
```diff
@@ -5,9 +5,12 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
+using System.Security.Cryptography;
```
🙄 VS

rynowak replied:
ugh. I'm going to fix this have to rebuild, and only because its the crypto namespace :(

SteveSandersonMS reviewed (approved):
Great!
